app/models/component.rb

```diff
@@ -16,27 +16,35 @@ class Component < Aggregate
              -> { where(usage: CONSTANTS::ENCRYPTION) }, class_name: 'Certificate', optional: true
 
   def self.to_service_metadata(event_id, published_at = Time.now)
-    matching_service_adapters = MsaComponent.includes(:enabled_signing_certificates,
-                                                      :encryption_certificate)
-    service_providers = SpComponent.includes(:enabled_signing_certificates,
-                                             :encryption_certificate)
+    matching_service_adapters = MsaComponent.all_components
+    service_providers = SpComponent.all_components
     components = matching_service_adapters + service_providers
 
     {
       published_at: published_at,
       event_id: event_id,
-      connected_services: components.map(&:services_to_metadata).flatten,
+      connected_services: components.map(&:services_to_metadata).flatten.compact,
       matching_service_adapters: matching_service_adapters.map(&:to_metadata),
       service_providers: service_providers.map(&:to_metadata)
     }
   end
 
+  def self.all_components
+    self.joins(:certificates)
+        .where(certificates: {
+          usage: [CONSTANTS::SIGNING, CONSTANTS::ENCRYPTION],
+          enabled: true,
+        }).distinct
+  end
+
   def services_to_metadata
     services.map do |service|
```

Comment:
When you request your components you probably want to do a `.includes(:service)` otherwise you're going to be making [N+1 requests](https://blog.appsignal.com/2018/04/24/active-record-performance-the-n+1-queries-antipattern.html).

As you seem to only want to do this request for Services that are attached to ServiceProviders you probably can change `components = matching_service_adapters + service_providers` above to `components = service_providers`

Reply:
👍 I am starting understand what we are making! 😛